engine/consensus/compliance/engine_test.go

```diff
@@ -14,74 +16,88 @@ import (
 	modulemock "github.com/onflow/flow-go/module/mock"
+	"github.com/onflow/flow-go/module/util"
 	"github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/network/mocknetwork"
 	"github.com/onflow/flow-go/utils/unittest"
 )
 
 func TestComplianceEngine(t *testing.T) {
-	suite.Run(t, new(ComplianceSuite))
+	suite.Run(t, new(EngineSuite))
 }
 
-type ComplianceSuite struct {
-	ComplianceCoreSuite
+// EngineSuite tests the compliance engine.
+type EngineSuite struct {
+	CommonSuite
 
+	ctx    irrecoverable.SignalerContext
+	cancel context.CancelFunc
+	errs   <-chan error
 	engine *Engine
 }
 
-func (cs *ComplianceSuite) SetupTest() {
-	cs.ComplianceCoreSuite.SetupTest()
+func (cs *EngineSuite) SetupTest() {
+	cs.CommonSuite.SetupTest()
+	cs.hotstuff.On("Start", mock.Anything)
+	cs.hotstuff.On("Ready", mock.Anything).Return(unittest.ClosedChannel()).Maybe()
+	cs.hotstuff.On("Done", mock.Anything).Return(unittest.ClosedChannel()).Maybe()
+
 	e, err := NewEngine(unittest.Logger(), cs.net, cs.me, cs.prov, cs.core)
 	require.NoError(cs.T(), err)
+	e.WithConsensus(cs.hotstuff)
 	cs.engine = e
 
-	ready := func() <-chan struct{} {
-		channel := make(chan struct{})
-		close(channel)
-		return channel
-	}()
+	cs.ctx, cs.cancel, cs.errs = irrecoverable.WithSignallerAndCancel(context.Background())
+	cs.engine.Start(cs.ctx)
 
-	cs.hotstuff.On("Start", mock.Anything)
-	cs.hotstuff.On("Ready", mock.Anything).Return(ready)
-	<-cs.engine.Ready()
+	unittest.AssertClosesBefore(cs.T(), cs.engine.Ready(), time.Second)
+}
+
+// TearDownTest stops the engine and checks there are no errors thrown to the SignallerContext.
+func (cs *EngineSuite) TearDownTest() {
+	cs.cancel()
+	<-cs.engine.Done()
+	select {
+	case err := <-cs.errs:
+		assert.NoError(cs.T(), err)
+	default:
+	}
 }
 
-// TestSendVote tests that single vote can be send and properly processed
-func (cs *ComplianceSuite) TestSendVote() {
+// TestSendVote tests that single vote can be sent and properly processed
+func (cs *EngineSuite) TestSendVote() {
 	// create parameters to send a vote
 	blockID := unittest.IdentifierFixture()
 	view := rand.Uint64()
 	sig := unittest.SignatureFixture()
 	recipientID := unittest.IdentifierFixture()
+	vote := &messages.BlockVote{
+		BlockID: blockID,
+		View:    view,
+		SigData: sig,
+	}
+
+	done := make(chan struct{})
+	*cs.con = *mocknetwork.NewConduit(cs.T())
+	cs.con.On("Unicast", vote, recipientID).
+		Run(func(_ mock.Arguments) { close(done) }).
+		Return(nil).
+		Once()
 
 	// submit the vote
 	err := cs.engine.SendVote(blockID, view, sig, recipientID)
 	require.NoError(cs.T(), err, "should pass send vote")
 
-	done := func() <-chan struct{} {
-		channel := make(chan struct{})
-		close(channel)
-		return channel
-	}()
-
-	cs.hotstuff.On("Done", mock.Anything).Return(done)
-
-	// The vote is transmitted asynchronously. We allow 10ms for the vote to be received:
-	<-time.After(10 * time.Millisecond)
+	// wait for vote to be sent
+	<-done
+	cs.cancel()
```

Comment:
any reason for cancelling manually? We already do this in `TearDown`

Reply:
You're right, no need to cancel within the test. Removing all these...